### PR TITLE
feat(Realtime Database): Clone from filtered dataset

### PR DESCRIPTION
**Pull request by davideast**
Addresses #244. 

Adds the ability to clone from a filtered dataset. Provides a checkbox to toggle the ability.

image

image

### CODE REVIEW

File: src/components/Database/DataViewer/CloneDialog.tsx [hunk 1]
```diff
@@ -63,19 +67,33 @@ const cloneKey = (key: string, ref: firebase.database.Reference) => {
...
+  let query =
+    queryParams == null ? null : applyQuery(realtimeRef, queryParams!);
+  let hasQuery = query !== null;
+
   const originalKey = realtimeRef.key!;
   const [newDestinationPath, setNewDestinationPath] = useState('');
 
   const [form, setForm] = useState<Record<string, string>>({});
   const [isLoading, setIsLoading] = useState(true);
+  const [cloneFiltered, setCloneFiltered] = useState(hasQuery);
```
Comment (tstirrat) [inline src/components/Database/DataViewer/CloneDialog.tsx:85]: isX/shouldX/hasX naming

File: src/components/Database/DataViewer/CloneDialog.tsx [hunk 2]
```diff
@@ -63,19 +67,33 @@ const cloneKey = (key: string, ref: firebase.database.Reference) => {
 
 export const CloneDialog = React.memo<Props>(function CloneDialog$({
   realtimeRef,
+  queryParams,
   onComplete,
 }) {
   ok(realtimeRef.parent, 'Cannot clone the root node!');
 
+  // queries only exist if queryParams are passed into the dialog
+  let query =
```
Comment (tstirrat) [inline src/components/Database/DataViewer/CloneDialog.tsx:76]: I think this is:

```js
let query = queryParams ?? applyQuery(realtimeRef, queryParams);
```

File: src/components/Database/DataViewer/common/fetch.ts [hunk 3]
```diff
@@ -139,12 +139,12 @@ function toObservable(query: firebase.database.Query) {
...
 }
 
-function applyQuery(
+export function applyQuery(
   ref: firebase.database.Reference,
   params: QueryParams
 ): firebase.database.Query {
   const { key, operator, value, limit } = params;
-  if (key && operator && value) {
+  if (key != undefined && operator != undefined && value != undefined) {
```
Comment (tstirrat) [inline src/components/Database/DataViewer/common/fetch.ts:147]: It's unclear here why `!=` is necessary... Perhaps a comment above to clarify that the != is intentional

File: src/components/Database/DataViewer/CloneDialog.tsx [hunk 4]
```diff
@@ -86,7 +104,14 @@ export const CloneDialog = React.memo<Props>(function CloneDialog$({
...
     loadData();
-  }, [setIsLoading, setForm, setNewDestinationPath, originalKey, realtimeRef]);
+  }, [
+    setIsLoading,
+    setForm,
+    setNewDestinationPath,
+    originalKey,
+    realtimeRef,
+    cloneFiltered,
+  ]);
```
Comment (tstirrat) [inline src/components/Database/DataViewer/CloneDialog.tsx:114]: needs `query` added here

File: src/components/Database/DataViewer/CloneDialog.tsx [hunk 5]
```diff
@@ -63,19 +67,33 @@ const cloneKey = (key: string, ref: firebase.database.Reference) => {
...
+  const [cloneFiltered, setCloneFiltered] = useState(hasQuery);
 
   useEffect(() => {
     const loadData = async () => {
-      const snapshot = await realtimeRef.once('value');
+      let snapshot = null;
+      // only apply the query if the user has selected the checkbox
+      // and the query can successfully be created from params
+      if (query != null && cloneFiltered) {
+        snapshot = await query.once('value');
```
Comment (tstirrat) [inline src/components/Database/DataViewer/CloneDialog.tsx:93]: need a test for this

File: src/components/Database/DataViewer/CloneDialog.tsx (hunk 2 above)
Comment (davideast) [reply src/components/Database/DataViewer/CloneDialog.tsx]: Removed applyQuery and now we just pass the query.

File: src/components/Database/DataViewer/common/fetch.ts (hunk 3 above)
Comment (davideast) [reply src/components/Database/DataViewer/common/fetch.ts]: Added the comment. tldr - it prevents bugs when the "value" needs to be false.

File: src/components/Database/DataViewer/CloneDialog.tsx (hunk 4 above)
Comment (davideast) [reply src/components/Database/DataViewer/CloneDialog.tsx:110]: done.

File: src/components/Database/DataViewer/CloneDialog.tsx (hunk 5 above)
Comment (davideast) [reply src/components/Database/DataViewer/CloneDialog.tsx:89]: Done.

File: src/components/Database/DataViewer/NodeActions.tsx [hunk 6]
```diff
@@ -110,6 +111,8 @@ export const NodeActions = React.memo<Props>(function NodeActions$({
     onDisplayTypeChange(getNextDisplayType(displayType));
   };
 
+  const potentialQuery = applyQuery(realtimeRef, queryParams || {});
```
Comment (tstirrat) [inline src/components/Database/DataViewer/NodeActions.tsx:114]: This query will be a new object each time I think, can you `useMemo` on it?

Comment (davideast) [reply src/components/Database/DataViewer/NodeActions.tsx]: Done.

**Pull request merged**